### DISCUSSION TITLE
Generate an Opaque Token inside a test

### DISCUSSION THREAD

**Discussion by LeCoupa** [adonisjs/core#2900] [Help] [answered]
Is there an easy way to create a token for a user we just created in a test (i.e. we do not have access to the request as I am inside Japa)?

I would like to create a helper in order to quickly create users and their associated token.

```javascript
const user = await User.create({
  email: faker.internet.email(),
  password: faker.internet.password(),
})

// Create a token for this user
// But this will not work as we do not have access to auth
await auth.use('api').attempt(user.email, user.password)
```

**Comment by arthurcvm** [comment]
Hi!
You can return the token in your helper, like you show.
Particularly I use a `http request`, but I believe that not is the best practice.
I create a test user in seeder and run a `group.before` in `test.group` to generate and get a token in response. Look:

```ts
let token = ''
group.before(async () => {
    const { body } = await supertest(`http://${String(process.env.HOST)}:${String(process.env.PORT)}`)
      .post('/login')
      .send({
        username: 'test_user',
        password: 'secret'
      })
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200)

      token = body.data.token.token
  })
```

**Comment by LeCoupa** [comment, marked as answer]
From the [Trello documentation](https://trello.com/c/NejT6yf6/45-july-planning-2021?utm_campaign=Monthly%20updates&utm_medium=email&utm_source=Revue%20newsletter) it seems that this feature is still in progress:

> Add helpers to the auth package to login users as a client (for testing)

**Reply by marcostaborda** [reply]
You can create a function inside a utils folder and call it when you want to use a tokenized user.

**Reply by marcostaborda** [reply]
```ts
import { UserFactory } from "Database/factories"
import { request } from "./request"

export const generateAuthUser = async () => { 
  const user = await UserFactory.merge({ password: 'secret' }).create()

  const { body } = await request.post('/auth').send({
    email: user.email,
    password: 'secret'
  })

  return { token: body.token, user } 
}
```
